File: fight_env/player/stats.py
from fight_env.inventory.armour import Armour, armour_types, ArmourTypes
from fight_env.inventory.shields import Shield, shields, Shields
from fight_env.inventory.weapons import Weapon, weapons, Weapons
from functools import wraps

from fight_env.player.refs.events import Events, Event
from fight_env.player.refs.tasks import TaskData

def lazy_recalc(method):
    @wraps(method)
    def wrapper(self, *args, **kwargs):
        if self._recalc_needed:
            self._update_values()
            self._recalc_needed = False
        return method(self, *args, **kwargs)

    return wrapper

def make_dirty(method):
    @wraps(method)
    def wrapper(self, *args, **kwargs):
        self._recalc_needed = True
        return method(self, *args, **kwargs)

    return wrapper


class Stats:
    def __init__(self,
                 name: str = "",
                 hp: int = 30,
                 stamina: int = 20):
        self.name: str = name
        self._base_hp: int = hp
        self._base_stamina: int = stamina
        self._hp: int = hp
        self._stamina: int = stamina

        self._armour: Armour = armour_types[ArmourTypes.NONE]
        self._shield: Shield = shields[Shields.NONE]
        self._weapon: Weapon = weapons[Weapons.NONE]

        self._recalc_needed: bool = True
        self._weight: int = 0

    @property
    def hp(self) -> int:
        return self._hp

    @property
    def stamina(self) -> int:
        return self._stamina

    @property
    def armour(self) -> Armour:
        return self._armour

    @armour.setter
    @make_dirty
    def armour(self, armour: ArmourTypes):
        self._armour = armour_types[armour]

    @property
    def shield(self) -> Shield:
        return self._shield

    @shield.setter
    @make_dirty
    def shield(self, shield: Shields):
        self._shield = shields[shield]

    @property
    def weapon(self) -> Weapon:
        return self._weapon

    @weapon.setter
    @make_dirty
    def weapon(self, weapon: Weapons):
        self._weapon = weapons[weapon]

    def _update_values(self):
        self._weight = self.armour.weight + self.shield.weight + self.weapon.weight
        # self._stamina_restore_value = BASE_STAMINA_RESTORE_VALUE_PER_FRAME
        self._base_stamina_expense = self._weight // 5

    @property
    def base_stamina_expense(self) -> int:
        return self._base_stamina_expense

    @property
    @lazy_recalc
    def max_stamina(self) -> int:
        return self._base_stamina - self._weight

    @property
    def max_hp(self) -> int:
        return self._base_hp

    @property
    @lazy_recalc
    def weight(self) -> int:
        return self._weight

    @property
    def damage_value(self) -> int:
        return self.weapon.base_damage

    @property
    def critical_damage_value(self) -> int:
        return self.weapon.critical_damage


def materialize_event(raw_event: Events, stats: Stats) -> Event:
    value_map = {
        Events.ATTACK: (Events.ATTACK, stats.damage_value),
        Events.CRITICAL_ATTACK: (Events.ATTACK, stats.critical_damage_value),
        Events.BLOCK: (Events.BLOCK, stats.shield.defense),
    }
    mapped_vent, value = value_map.get(raw_event, (raw_event, 0))
    return Event(mapped_vent, value)


def calc_stamina_cost_enter_task(task_data: TaskData, stats: Stats) -> int:
    return task_data.base_stamina_cost + stats.base_stamina_expense if task_data else 0


def calc_stamina_cost_frame(task_data: TaskData, stats: Stats) -> int:
    return task_data.base_stamina_cost_frame + stats.base_stamina_expense if task_data else 0

File: unity/Assets/FightCore/Stats.cs
// Port of fight_env/player/stats.py.
// The Python version recomputes lazily via dirty flags; here we just recompute
// eagerly whenever equipment changes -- behaviourally identical, simpler.
namespace FightCore
{
    public sealed class Stats
    {
        public string Name;

        private readonly int _baseHp;
        private readonly int _baseStamina;
        private int _hp;
        private int _stamina;

        private Armour _armour = Inventory.ArmourTable[ArmourTypes.NONE];
        private Shield _shield = Inventory.ShieldTable[Shields.NONE];
        private Weapon _weapon = Inventory.WeaponTable[Weapons.NONE];

        private int _weight;
        private int _baseStaminaExpense;

        public Stats(string name = "", int hp = 30, int stamina = 20)
        {
            Name = name;
            _baseHp = hp;
            _baseStamina = stamina;
            _hp = hp;
            _stamina = stamina;
            Recalc();
        }

        public int Hp => _hp;
        public int Stamina => _stamina;

        public Armour Armour => _armour;
        public Shield Shield => _shield;
        public Weapon Weapon => _weapon;

        public void SetArmour(ArmourTypes armour) { _armour = Inventory.ArmourTable[armour]; Recalc(); }
        public void SetShield(Shields shield) { _shield = Inventory.ShieldTable[shield]; Recalc(); }
        public void SetWeapon(Weapons weapon) { _weapon = Inventory.WeaponTable[weapon]; Recalc(); }

        private void Recalc()
        {
            _weight = _armour.Weight + _shield.Weight + _weapon.Weight;
            _baseStaminaExpense = _weight / 5; // Python floor-div; weight >= 0 so identical
        }

        public int BaseStaminaExpense => _baseStaminaExpense;
        public int Weight => _weight;
        public int MaxStamina => _baseStamina - _weight;
        public int MaxHp => _baseHp;
        public int DamageValue => _weapon.BaseDamage;
        public int CriticalDamageValue => _weapon.CriticalDamage;

        // fight_env/player/stats.py :: materialize_event
        public Event MaterializeEvent(Events raw)
        {
            switch (raw)
            {
                case Events.ATTACK:          return new Event(Events.ATTACK, DamageValue);
                case Events.CRITICAL_ATTACK: return new Event(Events.ATTACK, CriticalDamageValue);
                case Events.BLOCK:           return new Event(Events.BLOCK, _shield.Defense);
                default:                     return new Event(raw, 0);
            }
        }
    }
}